File: C++/Day12.cpp
#include "AH.h"

namespace Day12
{

	std::map<int64_t, int64_t> g_cache;

	int64_t numberOfSolns(const std::string & pattern, const std::vector<int> & is,
		const int pattern_idx, const int is_idx)
	{
		if (is_idx >= (int)is.size()) {
			// make sure we don't have any '#' left over
			if (pattern.find('#', pattern_idx) != std::string::npos) {
				return 0;
			}
			return 1; // nothing left to pack
		}

		// how much space do we need to pack the remaining blocks
 		int need = std::reduce(is.begin() + is_idx, is.end()) + is.size() - is_idx - 1;
		if (need > (int)pattern.size() - pattern_idx) {
			return 0;
		}

		int64_t waysToPack = 0;
		int64_t globaLookUp = pattern_idx * 100 + is_idx;
		if (g_cache.count(globaLookUp)) { // we've seen this situation before...
			//..so just return it
			return g_cache[globaLookUp];
		}
	  // try to fit the first blocks
	  int toFit = is[is_idx];
	  auto from = pattern.find_first_of("?#", pattern_idx);
	  if (from == std::string::npos) {
	  	// there are no ?s or #s left
	  	return 0;
	  }
	  for (size_t i = from;  i < pattern.size() - (toFit - 1); ++i) { // check for OBOE
		  // is there a . in the next toFit characters, if so can't write here
	  	if (pattern.find(".", i) < i + toFit) {
	  		continue;
	  	}

	  	// if there is a # before i we will miss it so there are no valid packings
	  	if (pattern.find('#', pattern_idx) < i) {
	  		break;
	  	}

	  	// we can write out block here so check the next char isn't a #
	  	// some optimisation here if is.size() == 1
	  	if (i + toFit < pattern.size()) {
	  		if (pattern.at(i + toFit) == '#') {
	  			continue;
	  		}
	  	}

	  	// at this point we can def write out so shift indices and recur
	  	auto next_pattern_idx = pattern.find_first_of("?#", i + toFit + 1);
	  	waysToPack += numberOfSolns(pattern, is, next_pattern_idx, is_idx + 1);
	  }

	  // cache this result so other branches can use it
	  g_cache[globaLookUp] = waysToPack;
	  return waysToPack;
	}

	int64_t parseInput(const std::string & s, const bool expand)
	{
		const auto ps = AH::Split(s, ' ');
		const auto qs = AH::Split(ps[1], ',');
		std::vector<int> partitions;
		for (auto q : qs) {
			partitions.push_back(std::stoi(q));
		}

		// reset cache before we start
		g_cache.clear();

		if (expand) {
			std::string str5 = ps[0];
			std::vector<int> parts5;

			for (auto i = 0; i < 4; ++ i) {
				str5 += '?'; // READ THE BLOODY RULES
				str5 += ps[0];
			}

			for (auto i = 0; i < 5; ++i) {
    		std::copy(partitions.begin(), partitions.end(), std::back_inserter(parts5));
			}

			return numberOfSolns(str5, parts5, 0, 0);
		}

		return numberOfSolns(ps[0], partitions, 0, 0);
	}

	int64_t Run(const std::string& filename)
	{
		const std::vector<std::string> ls = AH::ReadTextFile(filename);

		int part1 = 0;
		int64_t part2 = 0;
		int line = 1;
		for (auto l : ls) {
			part1 += parseInput(l, false);
			part2 += parseInput(l, true);
			line++;
		}

		AH::PrintSoln(12, part1, part2);

		return 0;
	}

}

File: Go/Day12/Day12.go
package Day12

import (
	"strconv"
	"strings"

	AH "AoC/adventhelper"
)

func numberOfSolns(pattern string, is []int, cache map[int]int) int {
	if 0 == len(is) {
		if AH.ContainsChar(pattern, '#') {
			return 0
		}
		return 1
	}

	need := 0
	for _, v := range is {
		need += (v + 1)
	}
	need -= 1
	if need > len(pattern) {
		return 0
	}

	hash := len(pattern)*100 + len(is)
	if v, ok := cache[hash]; ok {
		return v
	}

	toFit := is[0]
	from := strings.IndexAny(pattern, "?#")
	if from < 0 {
		return 0
	}

	waysToPack := 0
	for i := from; i < len(pattern)-toFit+1; i++ {
		t := strings.Index(pattern[i:], ".")
		if (t >= 0) && (t < toFit) {
			continue
		}

		t = strings.Index(pattern, "#")
		if (t >= 0) && (t < i) {
			break
		}

		if i+toFit < len(pattern) {
			if pattern[i+toFit] == '#' {
				continue
			}
		}

		next_idx := i + toFit + 1
		if next_idx > len(pattern) {
			next_idx = len(pattern)
		}
		waysToPack += numberOfSolns(pattern[next_idx:], is[1:], cache)
	}

	cache[hash] = waysToPack
	return waysToPack
}

func parseInput(s string, expand bool) int {
	ps := strings.Split(s, " ")
	qs := strings.Split(ps[1], ",")
	partitions := []int{}
	for _, q := range qs {
		i, _ := strconv.Atoi(q)
		partitions = append(partitions, i)
	}

	cache := make(map[int]int)

	ss := ps[0]
	pp := partitions
	if expand {
		for i := 0; i < 4; i++ {
			ss = ss + "?" + ps[0]
			pp = append(pp, partitions...)
		}
	}
	return numberOfSolns(ss, pp, cache)
}

func Run() {
	ls, _ := AH.ReadStrFile("../input/input12.txt")
	part1, part2 := 0, 0
	for _, l := range ls {
		part1 += parseInput(l, false)
		part2 += parseInput(l, true)
	}

	AH.PrintSoln(12, part1, part2)

	return
}
